<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Soham Ghosh" />

<meta name="date" content="2024-12-16" />

<title>Inflammatory Bowel Disease (IBD) Study</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">zinck-website</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="simulation.html">Simulations</a>
</li>
<li>
  <a href="Heatmaps.html">Heatmaps</a>
</li>
<li>
  <a href="CRC.html">CRC</a>
</li>
<li>
  <a href="IBD.html">IBD</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Inflammatory Bowel Disease (IBD) Study</h1>
<h4 class="author">Soham Ghosh</h4>
<h4 class="date">2024-12-16</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-12-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>zinck-website/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240617code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240617)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240617code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240617)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghoshstatszinckwebsitetreed4297f59f7c93bab5ff6585d3279bc9ad2cc9009targetblankd4297f5a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghoshstats/zinck-website/tree/d4297f59f7c93bab5ff6585d3279bc9ad2cc9009" target="_blank">d4297f5</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghoshstatszinckwebsitetreed4297f59f7c93bab5ff6585d3279bc9ad2cc9009targetblankd4297f5a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghoshstats/zinck-website/tree/d4297f59f7c93bab5ff6585d3279bc9ad2cc9009" target="_blank">d4297f5</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    analysis/.DS_Store

Untracked files:
    Untracked:  analysis/IBD.html

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/CRC.Rmd
    Deleted:    analysis/CRC.html
    Modified:   analysis/Heatmaps.Rmd
    Modified:   analysis/IBD.Rmd
    Modified:   analysis/_site.yml
    Modified:   analysis/index.Rmd
    Modified:   analysis/simulation.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/IBD.Rmd</code>) and HTML
(<code>docs/IBD.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghoshstats/zinck-website/d4297f59f7c93bab5ff6585d3279bc9ad2cc9009/docs/IBD.html" target="_blank">d4297f5</a>
</td>
<td>
ghoshstats
</td>
<td>
2024-12-12
</td>
<td>
Manually deploy website with existing HTML files
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghoshstats/zinck-website/d4046024117146dd0fe93cb71858798c88ab1fef/docs/IBD.html" target="_blank">d404602</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Updated IBD Analyses
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghoshstats/zinck-website/blob/a67317a066e6ca6dffb433788cd32b00ea837025/analysis/IBD.Rmd" target="_blank">a67317a</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Updated IBD Analyses
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghoshstats/zinck-website/ab6400d9cffa98ff7313a24230d9ecd128a1845a/docs/IBD.html" target="_blank">ab6400d</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Build and publish the website
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghoshstats/zinck-website/blob/a6c38f81ddf35b0d78bcc099268c13bc646997c5/analysis/IBD.Rmd" target="_blank">a6c38f8</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Add home, experiment, and simulation pages
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="ibd-data-analysis" class="section level2">
<h2>IBD Data Analysis</h2>
<p>Out of the ten uniformly processed 16S rRNA gene sequencing studies
of the IBD mucosal/stool microbiome (<a
href="https://github.com/biobakery/ibd_paper/tree/paper_publication"
class="uri">https://github.com/biobakery/ibd_paper/tree/paper_publication</a>),
we focus on five studies – RISK (<span
class="math inline">\(430\)</span> cases, <span
class="math inline">\(201\)</span> controls), CS PRISM (<span
class="math inline">\(359\)</span> cases, <span
class="math inline">\(38\)</span> controls), HMP2 (<span
class="math inline">\(59\)</span> cases, <span
class="math inline">\(22\)</span> controls), Pouchitis (<span
class="math inline">\(308\)</span> cases, <span
class="math inline">\(45\)</span> controls), and Mucosal IBD (<span
class="math inline">\(36\)</span> cases, <span
class="math inline">\(47\)</span> controls). Here cases indicate
patients with Ulcerative Colitis (UC) or Crohn’s Disease (CD). We
included all <span class="math inline">\(249\)</span> IBD genera in our
analyses.</p>
<pre class="r"><code>############################################################################################
################################## IBD data genus level ####################################

load(&quot;/Users/Patron/Documents/zinck research/genera.RData&quot;) ## Loading the meta IBD studies
combined_studies &lt;- as.data.frame(t(physeq_genera@otu_table))
study_names &lt;- physeq_genera@sam_data[[&quot;dataset_name&quot;]]

## RISK ##
risk_indices &lt;- which(study_names == &quot;RISK&quot;)
risk_otu &lt;- combined_studies[risk_indices, ]

## CS-PRISM ##
prism_indices &lt;- which(study_names == &quot;CS-PRISM&quot;)
prism_otu &lt;- combined_studies[prism_indices, ]

## HMP2 ##
HMP_indices &lt;- which(study_names == &quot;HMP2&quot;)
hmp_otu &lt;- combined_studies[HMP_indices, ]

## Pouchitis ##
pouchitis_indices &lt;- which(study_names == &quot;Pouchitis&quot;)
pouchitis_otu &lt;- combined_studies[pouchitis_indices, ]

## MucosalIBD ##
mi_indices &lt;- which(study_names == &quot;MucosalIBD&quot;)
mi_otu &lt;- combined_studies[mi_indices, ]

metadata &lt;- physeq_genera@sam_data

# Get the indices of the last occurrence of each unique subject_accession
latest_indices &lt;- sapply(unique(metadata$subject_accession), function(x) {
  max(which(metadata$subject_accession == x))
})

# Subset the metadata to keep only the latest samples
latest_metadata &lt;- metadata[latest_indices, ]

study_names &lt;- c(&quot;CS-PRISM&quot;,&quot;HMP2&quot;,&quot;MucosalIBD&quot;,&quot;Pouchitis&quot;,&quot;RISK&quot;)
meta_IBD &lt;- latest_metadata[latest_metadata$dataset_name %in% study_names,]

# Extract row names for each study
risk_row_names &lt;- row.names(latest_metadata[latest_metadata$dataset_name == &quot;RISK&quot;, ])
prism_row_names &lt;- row.names(latest_metadata[latest_metadata$dataset_name == &quot;CS-PRISM&quot;, ])
hmp_row_names &lt;- row.names(latest_metadata[latest_metadata$dataset_name == &quot;HMP2&quot;, ])
mi_row_names &lt;- row.names(latest_metadata[latest_metadata$dataset_name == &quot;MucosalIBD&quot;, ])
pouchitis_row_names &lt;- row.names(latest_metadata[latest_metadata$dataset_name == &quot;Pouchitis&quot;, ])

# Subset the OTU matrices and Y vectors based on these row names
risk_otu_latest &lt;- risk_otu[row.names(risk_otu) %in% risk_row_names, ]
prism_otu_latest &lt;- prism_otu[row.names(prism_otu) %in% prism_row_names, ]
hmp_otu_latest &lt;- hmp_otu[row.names(hmp_otu) %in% hmp_row_names, ]
mi_otu_latest &lt;- mi_otu[row.names(mi_otu) %in% mi_row_names, ]
pouchitis_otu_latest &lt;- pouchitis_otu[row.names(pouchitis_otu) %in% pouchitis_row_names, ]

combined_otu &lt;- rbind(prism_otu_latest,hmp_otu_latest,mi_otu_latest,
                      pouchitis_otu_latest,risk_otu_latest)
combined_otu &lt;- combined_otu[ match(rownames(meta_IBD), rownames(combined_otu)), ] 
## To make sure samples in OTU correspond to the samples in meta

IBD_resp &lt;- meta_IBD$disease
Y &lt;- ifelse(IBD_resp %in% c(&quot;CD&quot;, &quot;UC&quot;), 1, 0)

X &lt;- combined_otu</code></pre>
<p>We train the <code>zinck</code> model on <span
class="math inline">\(X\)</span> with the optimal number of clusters
(<span class="math inline">\(27\)</span>), and use the posterior
estimates of the latent parameters to generate the knockoff matrix.</p>
<pre class="r"><code>fitIBD &lt;- fit.zinck(as.matrix(X),num_clusters=27,method=&quot;ADVI&quot;,seed=1,
                    importance_resampling = TRUE,elbo_samples = 500,
                    boundary_correction = TRUE,prior_ZIGD = TRUE)
theta &lt;- fitIBD$theta
beta &lt;- fitIBD$beta
X_tilde &lt;- zinck::generateKnockoff(X,theta,beta,seed=1) ## Generating the knockoff copy</code></pre>
<p>Fitting the Random Forest model associating the augmented set of
covariates with the outcome of interest, we extract the Feature
Importance scores.</p>
<pre class="r"><code>filter_zinck &lt;- zinck.filter(as.matrix(X),as.matrix(X_tilde),as.factor(Y),
                             model=&quot;Random Forest&quot;,offset=0,seed=312)

selected_genera &lt;- filter_zinck$selected


## Importance scores ##
W &lt;- filter_zinck$W

## Threshold ##
T &lt;- filter_zinck$T

Xnorm &lt;- X / rowSums(X)

# Calculate the column sums for cases and controls
case_sums &lt;- colMeans(Xnorm[Y == 1, which(W&gt;=T)])
control_sums &lt;- colMeans(Xnorm[Y == 0, which(W&gt;=T)])

colors &lt;- ifelse(case_sums &gt; control_sums, &quot;red&quot;, &quot;blue&quot;)


# Create a data frame for plotting
data.genus &lt;- data.frame(
  impscores = sort(W[which(W &gt;= T)], decreasing = FALSE),
  name = factor(names_zinck, levels = names_zinck),
  y = seq(length(names_zinck)) * 0.9,
  color = colors
)


# Create the plot
plt.genus &lt;- ggplot(data.genus) +
  geom_col(aes(impscores, name, fill = color), width = 0.6) +
  scale_fill_identity() +
  theme_bw() +
  ylab(&quot;Genera&quot;) +
  xlab(&quot;Feature Statistic&quot;) +
  theme(
    axis.title.x = element_text(size = 22),
    axis.title.y = element_text(size = 22),
    axis.text.x = element_text(size = 18),
    axis.text.y = element_text(size = 18) 
  )
print(plt.genus)</code></pre>
<p><img src="figure/IBD.Rmd/imp-scores-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-imp-scores-1">
Past versions of imp-scores-1.png
</button>
</p>
<div id="fig-imp-scores-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ghoshstats/zinck-website/blob/d4046024117146dd0fe93cb71858798c88ab1fef/docs/figure/IBD.Rmd/imp-scores-1.png" target="_blank">d404602</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
</tr>
<tr>
<td>
<a href="https://github.com/ghoshstats/zinck-website/blob/ab6400d9cffa98ff7313a24230d9ecd128a1845a/docs/figure/IBD.Rmd/imp-scores-1.png" target="_blank">ab6400d</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that the red coloured bars indicate positive marginal
association between microbial relative abundance and the odds of IBD,
while blue indicate negative marginal association.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.3 (2022-03-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur/Monterey 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] kosel_0.0.1          phyloseq_1.38.0      rstan_2.21.8        
[4] StanHeaders_2.21.0-7 ggplot2_3.4.2        knockoff_0.3.6      
[7] reshape2_1.4.4       zinck_0.0.0.9000    

loaded via a namespace (and not attached):
  [1] nlme_3.1-162           bitops_1.0-7           matrixStats_0.63.0    
  [4] fs_1.6.2               rprojroot_2.0.3        GenomeInfoDb_1.30.1   
  [7] tools_4.1.3            bslib_0.5.0            vegan_2.6-4           
 [10] utf8_1.2.3             R6_2.5.1               mgcv_1.8-42           
 [13] DBI_1.1.3              BiocGenerics_0.40.0    colorspace_2.1-0      
 [16] permute_0.9-7          rhdf5filters_1.6.0     ade4_1.7-22           
 [19] withr_2.5.0            tidyselect_1.2.0       gridExtra_2.3         
 [22] prettyunits_1.1.1      processx_3.8.1         compiler_4.1.3        
 [25] git2r_0.32.0           glmnet_4.1-7           cli_3.6.1             
 [28] Biobase_2.54.0         labeling_0.4.2         sass_0.4.6            
 [31] scales_1.2.1           randomForest_4.7-1.1   callr_3.7.3           
 [34] stringr_1.5.0          digest_0.6.31          rmarkdown_2.22        
 [37] ordinalNet_2.12        XVector_0.34.0         pkgconfig_2.0.3       
 [40] htmltools_0.5.5        highr_0.10             fastmap_1.1.1         
 [43] rlang_1.1.1            rstudioapi_0.14        farver_2.1.1          
 [46] shape_1.4.6            jquerylib_0.1.4        generics_0.1.3        
 [49] jsonlite_1.8.5         dplyr_1.1.2            inline_0.3.19         
 [52] RCurl_1.98-1.12        magrittr_2.0.3         GenomeInfoDbData_1.2.7
 [55] loo_2.6.0              biomformat_1.22.0      Matrix_1.5-1          
 [58] Rcpp_1.0.10            munsell_0.5.0          S4Vectors_0.32.4      
 [61] Rhdf5lib_1.16.0        fansi_1.0.4            ape_5.7-1             
 [64] lifecycle_1.0.3        stringi_1.7.12         whisker_0.4.1         
 [67] yaml_2.3.7             MASS_7.3-60            zlibbioc_1.40.0       
 [70] rhdf5_2.38.1           pkgbuild_1.4.2         plyr_1.8.8            
 [73] grid_4.1.3             parallel_4.1.3         promises_1.2.0.1      
 [76] crayon_1.5.2           lattice_0.21-8         Biostrings_2.62.0     
 [79] splines_4.1.3          multtest_2.50.0        knitr_1.43            
 [82] ps_1.7.5               pillar_1.9.0           ranger_0.15.1         
 [85] igraph_1.4.2           codetools_0.2-19       stats4_4.1.3          
 [88] glue_1.6.2             evaluate_0.21          data.table_1.14.8     
 [91] RcppParallel_5.1.7     vctrs_0.6.5            httpuv_1.6.11         
 [94] foreach_1.5.2          gtable_0.3.3           cachem_1.0.8          
 [97] xfun_0.39              later_1.3.1            survival_3.5-5        
[100] tibble_3.2.1           iterators_1.0.14       IRanges_2.28.0        
[103] cluster_2.1.4          workflowr_1.7.1       </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
