<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Soham Ghosh" />

<meta name="date" content="2024-06-18" />

<title>Inflammatory Bowel Disease (IBD) Study</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">zinck-website</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Inflammatory Bowel Disease (IBD) Study</h1>
<h4 class="author">Soham Ghosh</h4>
<h4 class="date">2024-06-18</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-06-18
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>zinck-website/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240617code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240617)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240617code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240617)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghoshstatszinckwebsitetree6e66a4e2865fa53e2f40c14be091fe8ba20ce787targetblank6e66a4ea">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghoshstats/zinck-website/tree/6e66a4e2865fa53e2f40c14be091fe8ba20ce787" target="_blank">6e66a4e</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghoshstatszinckwebsitetree6e66a4e2865fa53e2f40c14be091fe8ba20ce787targetblank6e66a4ea"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghoshstats/zinck-website/tree/6e66a4e2865fa53e2f40c14be091fe8ba20ce787" target="_blank">6e66a4e</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/CRC.Rmd
    Untracked:  analysis/Heatmaps.Rmd
    Untracked:  analysis/IBD.Rmd
    Untracked:  analysis/simulation.Rmd

Unstaged changes:
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="ibd-data-analysis" class="section level2">
<h2>IBD Data Analysis</h2>
<p>Out of the ten uniformly processed 16S rRNA gene sequencing studies
of the IBD mucosal/stool microbiome (<a
href="https://github.com/biobakery/ibd_paper/tree/paper_publication"
class="uri">https://github.com/biobakery/ibd_paper/tree/paper_publication</a>),
we focus on five studies – RISK (430 cases, 201 controls), CS PRISM (359
cases, 38 controls), HMP2 (59 cases, 22 controls), Pouchitis (308 cases,
45 controls), and Mucosal IBD (36 cases, 47 controls). Here cases
indicate patients with Ulcerative Colitis (UC) or Crohn’s Disease (CD).
We included all <span class="math inline">\(249\)</span> IBD genera in
our analyses.</p>
<pre class="r"><code>library(zinck)
library(reshape2)
library(knockoff)
library(ggplot2)
library(rstan)</code></pre>
<pre><code>Loading required package: StanHeaders</code></pre>
<pre><code>rstan (Version 2.21.8, GitRev: 2e1f913d3ca3)</code></pre>
<pre><code>For execution on a local, multicore CPU with excess RAM we recommend calling
options(mc.cores = parallel::detectCores()).
To avoid recompilation of unchanged Stan programs, we recommend calling
rstan_options(auto_write = TRUE)</code></pre>
<pre class="r"><code>library(phyloseq)

################################################################################
###################### IBD data genus level ####################################

load(&quot;/Users/Patron/Documents/zinLDA research/genera.RData&quot;) ## Loading the meta IBD studies
combined_studies &lt;- as.data.frame(t(physeq_genera@otu_table))
study_names &lt;- physeq_genera@sam_data[[&quot;dataset_name&quot;]]

#### Study : RISK ####
risk_indices &lt;- which(study_names == &quot;RISK&quot;)
risk_otu &lt;- combined_studies[risk_indices, ]
IBD_resp &lt;- physeq_genera@sam_data[[&quot;disease&quot;]][risk_indices]
risk_Y &lt;- ifelse(IBD_resp %in% c(&quot;CD&quot;, &quot;UC&quot;), 1, 0) ## Labelling &quot;CD&quot; or &quot;UC&quot; as 1, rest as 0

#### Study : CS-PRISM ####
prism_indices &lt;- which(study_names == &quot;CS-PRISM&quot;)
prism_otu &lt;- combined_studies[prism_indices, ]
IBD_resp &lt;- physeq_genera@sam_data[[&quot;disease&quot;]][prism_indices]
prism_Y &lt;- ifelse(IBD_resp %in% c(&quot;CD&quot;, &quot;UC&quot;), 1, 0) 

#### Study : HMP2 ######

hmp_indices &lt;- which(study_names == &quot;HMP2&quot;)
hmp_otu &lt;- combined_studies[hmp_indices, ]
IBD_resp &lt;- physeq_genera@sam_data[[&quot;disease&quot;]][hmp_indices]
hmp_Y &lt;- ifelse(IBD_resp %in% c(&quot;CD&quot;, &quot;UC&quot;), 1, 0) 

##### Study : MucosalIBD #####

mi_indices &lt;- which(study_names == &quot;MucosalIBD&quot;)
mi_otu &lt;- combined_studies[mi_indices, ]
IBD_resp &lt;- physeq_genera@sam_data[[&quot;disease&quot;]][mi_indices]
mi_Y &lt;- ifelse(IBD_resp %in% c(&quot;CD&quot;, &quot;UC&quot;), 1, 0) 

##### Study : Pouchitis #####

pouchitis_indices &lt;- which(study_names == &quot;Pouchitis&quot;)
pouchitis_otu &lt;- combined_studies[pouchitis_indices, ]
IBD_resp &lt;- physeq_genera@sam_data[[&quot;disease&quot;]][pouchitis_indices]
pouchitis_Y &lt;- ifelse(IBD_resp %in% c(&quot;CD&quot;, &quot;UC&quot;), 1, 0) 


######### Combining all the 5 studies together #############
X &lt;- rbind(risk_otu,prism_otu,hmp_otu,mi_otu,pouchitis_otu)
Y &lt;- c(risk_Y,prism_Y,hmp_Y,mi_Y,pouchitis_Y)</code></pre>
<p>We train the zinck model on <span class="math inline">\(X\)</span>
with the optimal number of clusters (27), and use the posterior
estimates of the latent parameters to generate the knockoff matrix.</p>
<pre class="r"><code>zinck_code &lt;- &quot;data {
  int&lt;lower=1&gt; K; // num topics
  int&lt;lower=1&gt; V; // num words
  int&lt;lower=0&gt; D; // num docs
  int&lt;lower=0&gt; n[D, V]; // word counts for each doc

  // hyperparameters
  vector&lt;lower=0, upper=1&gt;[V] delta;
}

parameters {
  simplex[K] theta[D]; // topic mixtures
  vector&lt;lower=0, upper=1&gt;[V] zeta[K]; // zero-inflated betas
  vector&lt;lower=0&gt;[V] gamma1[K];
  vector&lt;lower=0&gt;[V] gamma2[K];
  vector&lt;lower=0&gt;[K] alpha;
}

transformed parameters {
  vector[V] beta[K];

  // Efficiently compute beta using vectorized operations
  for (k in 1:K) {
    vector[V] cum_log1m;
    cum_log1m[1:(V - 1)] = cumulative_sum(log1m(zeta[k, 1:(V - 1)]));
    cum_log1m[V] = 0;
    beta[k] = zeta[k] .* exp(cum_log1m);
    beta[k] = beta[k] / sum(beta[k]);
  }
}


model {
  for (k in 1:K) {
    alpha[k] ~ gamma(100,100);  // Change these hyperparameters as needed
  }
  for (d in 1:D) {
    theta[d] ~ dirichlet(alpha);
  }
  for (k in 1:K) {
    for (m in 1:V) {
        gamma1[k,m] ~ gamma(1,1);
        gamma2[k,m] ~ gamma(1,1);
    }
  }

  // Zero-inflated beta likelihood and data likelihood
  for (k in 1:K) {
    for (m in 1:V) {
      real lp_non_zero = bernoulli_lpmf(0 | delta[m]) + beta_lpdf(zeta[k, m] | gamma1[k, m], gamma2[k, m]);
      real lp_zero = bernoulli_lpmf(1 | delta[m]);
      target += log_sum_exp(lp_non_zero, lp_zero);
    }
  }

  // Compute the eta values and data likelihood more efficiently
  for (d in 1:D) {
    vector[V] eta = theta[d, 1] * beta[1];
    for (k in 2:K) {
      eta += theta[d, k] * beta[k];
    }
    eta = eta / sum(eta);
    n[d] ~ multinomial(eta);
  }
}
&quot;

stan.model = stan_model(model_code = zinck_code)</code></pre>
<pre><code>Trying to compile a simple C file</code></pre>
<pre><code>Running /Library/Frameworks/R.framework/Resources/bin/R CMD SHLIB foo.c
clang -mmacosx-version-min=10.13 -I&quot;/Library/Frameworks/R.framework/Resources/include&quot; -DNDEBUG   -I&quot;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/Rcpp/include/&quot;  -I&quot;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppEigen/include/&quot;  -I&quot;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppEigen/include/unsupported&quot;  -I&quot;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/BH/include&quot; -I&quot;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/StanHeaders/include/src/&quot;  -I&quot;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/StanHeaders/include/&quot;  -I&quot;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppParallel/include/&quot;  -I&quot;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/rstan/include&quot; -DEIGEN_NO_DEBUG  -DBOOST_DISABLE_ASSERTS  -DBOOST_PENDING_INTEGER_LOG2_HPP  -DSTAN_THREADS  -DBOOST_NO_AUTO_PTR  -include &#39;/Library/Frameworks/R.framework/Versions/4.1/Resources/library/StanHeaders/include/stan/math/prim/mat/fun/Eigen.hpp&#39;  -D_REENTRANT -DRCPP_PARALLEL_USE_TBB=1   -I/usr/local/include   -fPIC  -Wall -g -O2  -c foo.c -o foo.o
In file included from &lt;built-in&gt;:1:
In file included from /Library/Frameworks/R.framework/Versions/4.1/Resources/library/StanHeaders/include/stan/math/prim/mat/fun/Eigen.hpp:13:
In file included from /Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppEigen/include/Eigen/Dense:1:
In file included from /Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppEigen/include/Eigen/Core:88:
/Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppEigen/include/Eigen/src/Core/util/Macros.h:628:1: error: unknown type name &#39;namespace&#39;
namespace Eigen {
^
/Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppEigen/include/Eigen/src/Core/util/Macros.h:628:16: error: expected &#39;;&#39; after top level declarator
namespace Eigen {
               ^
               ;
In file included from &lt;built-in&gt;:1:
In file included from /Library/Frameworks/R.framework/Versions/4.1/Resources/library/StanHeaders/include/stan/math/prim/mat/fun/Eigen.hpp:13:
In file included from /Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppEigen/include/Eigen/Dense:1:
/Library/Frameworks/R.framework/Versions/4.1/Resources/library/RcppEigen/include/Eigen/Core:96:10: fatal error: &#39;complex&#39; file not found
#include &lt;complex&gt;
         ^~~~~~~~~
3 errors generated.
make: *** [foo.o] Error 1</code></pre>
<pre class="r"><code>X &lt;- X[,order(decreasing=T,colSums(X,na.rm=T),apply(X,2L,paste,collapse=&#39;&#39;))] ## ordering the columns w/ decreasing abundance

dlt &lt;- rep(0,ncol(X)) ## Initializing the deltas with the sparsity of each column

for(t in (1:ncol(X)))
{
  dlt[t] &lt;- 1-mean(X[,t]&gt;0)
  if(dlt[t]==0)
  {
    dlt[t] = dlt[t]+0.01
  }
  if (dlt[t]==1)
  {
    dlt[t] = dlt[t]-0.01
  }
  
}

zinLDA_stan_data &lt;- list(
  K = 27,
  V = ncol(X),
  D = nrow(X),
  n = X,
  delta = dlt
)

## Fitting the zinck model ##
set.seed(1)
fitIBD &lt;- vb(stan.model, data=zinLDA_stan_data, algorithm=&quot;meanfield&quot;, importance_resampling=TRUE, iter=10000,tol_rel_obj=0.01,elbo_samples=500)</code></pre>
<pre><code>Chain 1: ------------------------------------------------------------
Chain 1: EXPERIMENTAL ALGORITHM:
Chain 1:   This procedure has not been thoroughly tested and may be unstable
Chain 1:   or buggy. The interface is subject to change.
Chain 1: ------------------------------------------------------------
Chain 1: 
Chain 1: 
Chain 1: 
Chain 1: Gradient evaluation took 0.7248 seconds
Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 7248 seconds.
Chain 1: Adjust your expectations accordingly!
Chain 1: 
Chain 1: 
Chain 1: Begin eta adaptation.
Chain 1: Iteration:   1 / 250 [  0%]  (Adaptation)
Chain 1: Iteration:  50 / 250 [ 20%]  (Adaptation)
Chain 1: Iteration: 100 / 250 [ 40%]  (Adaptation)
Chain 1: Iteration: 150 / 250 [ 60%]  (Adaptation)
Chain 1: Iteration: 200 / 250 [ 80%]  (Adaptation)
Chain 1: Iteration: 250 / 250 [100%]  (Adaptation)
Chain 1: Success! Found best value [eta = 0.1].
Chain 1: 
Chain 1: Begin stochastic gradient ascent.
Chain 1:   iter             ELBO   delta_ELBO_mean   delta_ELBO_med   notes 
Chain 1:    100   -126700447.982             1.000            1.000
Chain 1:    200    -77685031.927             0.815            1.000
Chain 1:    300    -59307492.121             0.647            0.631
Chain 1:    400    -50860990.107             0.527            0.631
Chain 1:    500    -45537078.584             0.445            0.310
Chain 1:    600    -41685839.741             0.386            0.310
Chain 1:    700    -38580134.819             0.342            0.166
Chain 1:    800    -36053383.596             0.308            0.166
Chain 1:    900    -33979738.726             0.281            0.117
Chain 1:   1000    -32254605.026             0.258            0.117
Chain 1:   1100    -30792620.765             0.163            0.092
Chain 1:   1200    -29559706.949             0.104            0.081
Chain 1:   1300    -28474935.327             0.077            0.070
Chain 1:   1400    -27530699.160             0.064            0.061
Chain 1:   1500    -26698219.449             0.055            0.053
Chain 1:   1600    -25944401.570             0.049            0.047
Chain 1:   1700    -25287308.192             0.043            0.042
Chain 1:   1800    -24690624.955             0.039            0.038
Chain 1:   1900    -24160947.818             0.035            0.034
Chain 1:   2000    -23683162.157             0.031            0.031
Chain 1:   2100    -23249587.266             0.029            0.029
Chain 1:   2200    -22863841.826             0.026            0.026
Chain 1:   2300    -22511759.353             0.024            0.024
Chain 1:   2400    -22195116.004             0.022            0.022
Chain 1:   2500    -21911818.977             0.020            0.020
Chain 1:   2600    -21658464.199             0.018            0.019
Chain 1:   2700    -21427457.111             0.017            0.017
Chain 1:   2800    -21217341.358             0.015            0.016
Chain 1:   2900    -21028919.131             0.014            0.014
Chain 1:   3000    -20858398.183             0.013            0.013
Chain 1:   3100    -20704895.615             0.012            0.012
Chain 1:   3200    -20558554.504             0.011            0.011
Chain 1:   3300    -20420888.342             0.010            0.010   MEAN ELBO CONVERGED   MEDIAN ELBO CONVERGED
Chain 1: 
Chain 1: Drawing a sample of size 1000 from the approximate posterior... 
Chain 1: COMPLETED.</code></pre>
<pre><code>Warning: Pareto k diagnostic value is Inf. Resampling is disabled. Decreasing
tol_rel_obj may help if variational algorithm has terminated prematurely.
Otherwise consider using sampling instead.</code></pre>
<pre class="r"><code>theta &lt;- fitIBD@sim[[&quot;est&quot;]][[&quot;theta&quot;]]
beta &lt;- fitIBD@sim[[&quot;est&quot;]][[&quot;beta&quot;]]
X_tilde &lt;- zinck::generateKnockoff(X,theta,beta,seed=1) ## Generating the kncokoff copy</code></pre>
<p>Fitting the Random Forest model associating the augmented set of
covariates with the outcome of interest, we extract the Feature
Importance scores.</p>
<pre class="r"><code>set.seed(5)
W &lt;- stat.random_forest(X,X_tilde,Y)
T &lt;- knockoff.threshold(W,fdr=0.1,offset = 0) ## This is the knockoff filter threshold
print(which(W&gt;=T))</code></pre>
<pre><code> [1]  8 16 22 24 28 32 37 61 65 69</code></pre>
<pre class="r"><code>names &lt;- colnames(X[,which(W&gt;=T)]) ## Extracting the names of the important genera
data.genus &lt;- data.frame(
  impscores = sort(W[which(W&gt;=T)], decreasing=FALSE) , 
  name = factor(names, levels = names),
  y = seq(length(names)) * 0.9
)

plt.genus &lt;- ggplot(data.genus) +
  geom_col(aes(impscores, name), fill = &quot;black&quot;, width = 0.6)+theme_bw()+ylab(&quot;Genera&quot;)+xlab(&quot;Feature Statistic&quot;)

plt.genus</code></pre>
<p><img src="figure/IBD.Rmd/imp-scores-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.3 (2022-03-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur/Monterey 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] phyloseq_1.38.0      rstan_2.21.8         StanHeaders_2.21.0-7
[4] ggplot2_3.4.2        knockoff_0.3.6       reshape2_1.4.4      
[7] zinck_0.0.0.9000     workflowr_1.7.1     

loaded via a namespace (and not attached):
  [1] nlme_3.1-162           bitops_1.0-7           matrixStats_0.63.0    
  [4] fs_1.6.2               httr_1.4.6             rprojroot_2.0.3       
  [7] GenomeInfoDb_1.30.1    tools_4.1.3            bslib_0.5.0           
 [10] vegan_2.6-4            utf8_1.2.3             R6_2.5.1              
 [13] mgcv_1.8-42            DBI_1.1.3              BiocGenerics_0.40.0   
 [16] colorspace_2.1-0       permute_0.9-7          rhdf5filters_1.6.0    
 [19] ade4_1.7-22            withr_2.5.0            tidyselect_1.2.0      
 [22] gridExtra_2.3          prettyunits_1.1.1      processx_3.8.1        
 [25] compiler_4.1.3         git2r_0.32.0           glmnet_4.1-7          
 [28] cli_3.6.1              Biobase_2.54.0         labeling_0.4.2        
 [31] sass_0.4.6             scales_1.2.1           randomForest_4.7-1.1  
 [34] callr_3.7.3            stringr_1.5.0          digest_0.6.31         
 [37] rmarkdown_2.22         XVector_0.34.0         pkgconfig_2.0.3       
 [40] htmltools_0.5.5        highr_0.10             fastmap_1.1.1         
 [43] rlang_1.1.1            rstudioapi_0.14        farver_2.1.1          
 [46] shape_1.4.6            jquerylib_0.1.4        generics_0.1.3        
 [49] jsonlite_1.8.5         dplyr_1.1.2            inline_0.3.19         
 [52] RCurl_1.98-1.12        magrittr_2.0.3         GenomeInfoDbData_1.2.7
 [55] loo_2.6.0              biomformat_1.22.0      Matrix_1.5-1          
 [58] Rhdf5lib_1.16.0        Rcpp_1.0.10            munsell_0.5.0         
 [61] S4Vectors_0.32.4       fansi_1.0.4            ape_5.7-1             
 [64] lifecycle_1.0.3        stringi_1.7.12         whisker_0.4.1         
 [67] yaml_2.3.7             MASS_7.3-60            zlibbioc_1.40.0       
 [70] rhdf5_2.38.1           pkgbuild_1.4.2         plyr_1.8.8            
 [73] grid_4.1.3             parallel_4.1.3         promises_1.2.0.1      
 [76] crayon_1.5.2           lattice_0.21-8         Biostrings_2.62.0     
 [79] splines_4.1.3          multtest_2.50.0        knitr_1.43            
 [82] ps_1.7.5               pillar_1.9.0           ranger_0.15.1         
 [85] igraph_1.4.2           codetools_0.2-19       stats4_4.1.3          
 [88] glue_1.6.2             evaluate_0.21          getPass_0.2-2         
 [91] data.table_1.14.8      RcppParallel_5.1.7     vctrs_0.6.5           
 [94] httpuv_1.6.11          foreach_1.5.2          gtable_0.3.3          
 [97] cachem_1.0.8           xfun_0.39              later_1.3.1           
[100] survival_3.5-5         tibble_3.2.1           iterators_1.0.14      
[103] IRanges_2.28.0         cluster_2.1.4         </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
