<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Soham Ghosh" />


<title>Home</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">zinck-website</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="simulation.html">Simulations</a>
</li>
<li>
  <a href="Heatmaps.html">Heatmaps</a>
</li>
<li>
  <a href="CRC.html">CRC</a>
</li>
<li>
  <a href="IBD.html">IBD</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Home</h1>
<h4 class="author">Soham Ghosh</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-12-12
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 1
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>zinck-website/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghoshstatszinckwebsitetreed4297f59f7c93bab5ff6585d3279bc9ad2cc9009targetblankd4297f5a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghoshstats/zinck-website/tree/d4297f59f7c93bab5ff6585d3279bc9ad2cc9009" target="_blank">d4297f5</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghoshstatszinckwebsitetreed4297f59f7c93bab5ff6585d3279bc9ad2cc9009targetblankd4297f5a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghoshstats/zinck-website/tree/d4297f59f7c93bab5ff6585d3279bc9ad2cc9009" target="_blank">d4297f5</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    analysis/.DS_Store

Untracked files:
    Untracked:  analysis/IBD.html

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/CRC.Rmd
    Deleted:    analysis/CRC.html
    Modified:   analysis/Heatmaps.Rmd
    Modified:   analysis/IBD.Rmd
    Modified:   analysis/_site.yml
    Modified:   analysis/index.Rmd
    Modified:   analysis/simulation.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/index.Rmd</code>) and HTML
(<code>docs/index.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghoshstats/zinck-website/ab6400d9cffa98ff7313a24230d9ecd128a1845a/docs/index.html" target="_blank">ab6400d</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Build and publish the website
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghoshstats/zinck-website/blob/a6c38f81ddf35b0d78bcc099268c13bc646997c5/analysis/index.Rmd" target="_blank">a6c38f8</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Add home, experiment, and simulation pages
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghoshstats/zinck-website/blob/6e66a4e2865fa53e2f40c14be091fe8ba20ce787/analysis/index.Rmd" target="_blank">6e66a4e</a>
</td>
<td>
Patron
</td>
<td>
2024-06-17
</td>
<td>
Initial commit of Workflowr project
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="welcome-to-the-zinck-research-website" class="section level1">
<h1>Welcome to the Zinck Research Website!</h1>
<p>This website documents the research and experiments related to the
<code>zinck</code> package.</p>
<div id="description" class="section level2">
<h2>Description</h2>
<p>Zinck exploits a zero-inflated variant of the Latent Dirichlet
Allocation (LDA) model to generate valid knockoffs that capture the key
characteristics of microbiome data - mainly its compositional nature and
high sparsity. It exhibits the properties of simultaneous variable
selection and FDR control to identify microbial biomarkers. This package
provides an implementation of zinck, which is trained either using the
Automatic Differentiation Variational Inference (ADVI) algorithm or
using a collapsed Gibbs sampler, facilitating variable selection for
both continuous as well as binary outcomes.</p>
</div>
<div id="the-hierarchical-structure-of-zinck" class="section level2">
<h2>The hierarchical structure of zinck</h2>
<p>Before going into the structure of zinck, we define the following
parameters:</p>
<ol style="list-style-type: decimal">
<li><p>Number of biological samples: <span
class="math inline">\(D\)</span>, number of microbial features: <span
class="math inline">\(p\)</span>, number of latent clusters <span
class="math inline">\(K\)</span> and the sequencing depth of the <span
class="math inline">\(d^{th}\)</span> sample is given by <span
class="math inline">\(N_d\)</span>, where <span
class="math inline">\(d=1,\ldots,D\)</span>.</p></li>
<li><p><span class="math inline">\(\boldsymbol{\theta}_{d} =
(\theta_{d1},\ldots,\theta_{dK})&#39;\)</span> is the vector of cluster
mixing probabilities for the <span class="math inline">\(d^{th}\)</span>
biological sample, where <span class="math inline">\(d \in
1,2,\ldots,D\)</span></p></li>
<li><p><span class="math inline">\(\boldsymbol{\beta}_{k} =
(\beta_{k1},\ldots,\beta_{k(p-1)})&#39;\)</span> and <span
class="math inline">\(\boldsymbol{\beta}_{kp}=1-\sum_{i=1}^{p-1}\beta_{ki}\)</span>
is the vector of feature proportions for each cluster <span
class="math inline">\(k=1,2,\ldots,K\)</span>.</p></li>
<li><p><span
class="math inline">\(\mathbf{z}_{d}=\left(z_{d1},\ldots,z_{dN_{d}}\right)&#39;\)</span>
is the vector of cluster assignments for the <span
class="math inline">\(d^{th}\)</span> biological sample. For instance,
<span class="math inline">\(z_{dn}=k\)</span> implies that the <span
class="math inline">\(n^{th}\)</span> sequencing read in the <span
class="math inline">\(d^{th}\)</span> sample belongs to the <span
class="math inline">\(k^{th}\)</span> cluster.</p></li>
<li><p><span
class="math inline">\(\mathbf{w}_{d}=\left(w_{d1},\ldots,w_{dN_{d}}\right)&#39;\)</span>
is the vector of the features drawn for each sequencing read for the
<span class="math inline">\(d^{th}\)</span> biological sample.</p></li>
<li><p><span
class="math inline">\(\boldsymbol{\alpha}=(\alpha,\ldots,\alpha)^{K
\times 1}\)</span>: symmetric hyperparameter of the Dirichlet prior of
<span class="math inline">\(\boldsymbol{\theta}_{d}\)</span>.</p></li>
<li><p><span
class="math inline">\(\boldsymbol{\pi}_{k}=(\pi_{k1},\ldots,\pi_{k(p-1)})&#39;\)</span>:
hyperparameter of the ZIGD distribution specifying the probability of
being a structural zero for the <span
class="math inline">\(j^{th}\)</span> feature in the <span
class="math inline">\(k^{th}\)</span> subcommunity, where <span
class="math inline">\(j=1,\ldots,p\)</span> and <span
class="math inline">\(k=1,\ldots,K\)</span>.</p></li>
<li><p><span class="math inline">\(\mathbf{a}=(a,\ldots,a)^{(p-1) \times
1}\)</span> and <span class="math inline">\(\mathbf{b}=(b, \ldots,
b)^{(p-1) \times 1}\)</span> are the symmetric hyperparameter vectors on
the ZIGD of <span
class="math inline">\(\boldsymbol{\beta}_{k}\)</span>.</p></li>
</ol>
<p>It is to be noted that ZIGD here refers to the zero-inflated
Generalized Dirichlet distribution.</p>
<p>zinck is a probabilistic hierarchical model with the following
specification:</p>
<p><span class="math display">\[
\begin{aligned}
    w_{dn}|z_{dn},\boldsymbol{\beta}_{z_{dn}} &amp; \sim
\text{Multinomial}(\boldsymbol{\beta}_{z_{dn}}) \\
    \boldsymbol{\beta}_{z_{dn}}|\boldsymbol{\pi},\mathbf{a},\mathbf{b}
&amp; \sim \text{ZIGD}\left(\pi_{z_{dn}},a,b\right) \\
    z_{dn}|\theta_{d} &amp; \sim \text{Multinomial}(\theta_{d}) \\
    \boldsymbol{\theta}_{d} &amp; \sim
\text{Dirichlet}(\boldsymbol{\alpha})
\end{aligned}
\]</span></p>
<p>We denote the elements of the observed sample taxa matrix <span
class="math inline">\(\mathbf{X}^{D \times p}\)</span> by <span
class="math inline">\((x_{dj})\)</span> as the observed read count of
the <span class="math inline">\(j^{th}\)</span> taxon for the <span
class="math inline">\(d^{th}\)</span> subject: <span
class="math display">\[x_{dj}=\sum_{n=1}^{N_d}
\mathbb{1}_{\{w_{dn}=j\}}\]</span></p>
</div>
<div id="generating-knockoffs-using-zinck" class="section level2">
<h2>Generating knockoffs using zinck</h2>
<p>We can exploit the knockoff generative model to learn the structure
of the sample-feature count matrix <span
class="math inline">\(\mathbf{X}\)</span> and generate a valid knockoff
copy, which is further used for FDR-controlled feature selection. We fit
the augmented LDA model to the microbiome count data matrix <span
class="math inline">\(\mathbf{X}\)</span>. The latent parameters of the
model namely, <span class="math inline">\(\theta_{d}\)</span> and <span
class="math inline">\(\beta_{k}\)</span> are learnt by approximating
their joint posterior distribution via the Automatic Differentiation
Variational Inference (ADVI) algorithm or by drawing MCMC samples using
a Collapsed Gibbs sampler. We use the learnt parameters <span
class="math inline">\(\tilde{\boldsymbol{\theta}}_d,
\tilde{\boldsymbol{\beta}}_k\)</span>, for <span
class="math inline">\(d=1,2,\ldots D\)</span> and <span
class="math inline">\(k=1,2,\ldots,K\)</span> to generate a knockoff
copy. For each sample <span class="math inline">\(d\)</span>, we first
sample a cluster allocation <span class="math inline">\(z_{dn}\)</span>
from <span
class="math inline">\(\text{Multinomial}(1,\tilde{\boldsymbol{\theta}}_d)\)</span>
and then we sample a feature <span class="math inline">\(w_{dn}\)</span>
from the selected cluster <span class="math inline">\(z_{dn}\)</span>,
that is, <span class="math inline">\(w_{dn} \sim
\text{Multinomial}(1,\tilde{\boldsymbol{\beta}}_{z_{dn}})\)</span>.
Finally, we form the knockoff matrix <span
class="math inline">\(\tilde{\mathbf{X}}^{D \times p} =
\{\tilde{x}_{dj}\}\)</span> by cumulating all the taxa read counts per
subjects as illustrated previously.</p>
</div>
<div id="pages" class="section level2">
<h2>Pages</h2>
<ul>
<li><a href="Heatmaps.html">Heatmaps</a></li>
<li><a href="simulation.html">Simulation Experiments</a></li>
<li><a href="CRC.html">CRC Analysis</a></li>
<li><a href="IBD.html">IBD Analysis</a></li>
</ul>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
