<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Soham Ghosh" />

<meta name="date" content="2024-12-11" />

<title>Simulations</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">zinck-website</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="simulation.html">Simulations</a>
</li>
<li>
  <a href="Heatmaps.html">Heatmaps</a>
</li>
<li>
  <a href="CRC.html">CRC</a>
</li>
<li>
  <a href="IBD.html">IBD</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Simulations</h1>
<h4 class="author">Soham Ghosh</h4>
<h4 class="date">2024-12-11</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-12-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>zinck-website/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240617code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240617)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240617code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240617)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomghoshstatszinckwebsitetreed4297f59f7c93bab5ff6585d3279bc9ad2cc9009targetblankd4297f5a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/ghoshstats/zinck-website/tree/d4297f59f7c93bab5ff6585d3279bc9ad2cc9009" target="_blank">d4297f5</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomghoshstatszinckwebsitetreed4297f59f7c93bab5ff6585d3279bc9ad2cc9009targetblankd4297f5a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/ghoshstats/zinck-website/tree/d4297f59f7c93bab5ff6585d3279bc9ad2cc9009" target="_blank">d4297f5</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    analysis/.DS_Store

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/CRC.Rmd
    Deleted:    analysis/CRC.html
    Modified:   analysis/Heatmaps.Rmd
    Modified:   analysis/IBD.Rmd
    Modified:   analysis/_site.yml
    Modified:   analysis/index.Rmd
    Modified:   analysis/simulation.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/simulation.Rmd</code>) and HTML
(<code>docs/simulation.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghoshstats/zinck-website/d4297f59f7c93bab5ff6585d3279bc9ad2cc9009/docs/simulation.html" target="_blank">d4297f5</a>
</td>
<td>
ghoshstats
</td>
<td>
2024-12-12
</td>
<td>
Manually deploy website with existing HTML files
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghoshstats/zinck-website/blob/d39920f0e7ace7e77d041caafe3f4854c4d7851f/analysis/simulation.Rmd" target="_blank">d39920f</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Changed the simulation settings
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghoshstats/zinck-website/d39920f0e7ace7e77d041caafe3f4854c4d7851f/docs/simulation.html" target="_blank">d39920f</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Changed the simulation settings
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ghoshstats/zinck-website/ab6400d9cffa98ff7313a24230d9ecd128a1845a/docs/simulation.html" target="_blank">ab6400d</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Build and publish the website
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ghoshstats/zinck-website/blob/a6c38f81ddf35b0d78bcc099268c13bc646997c5/analysis/simulation.Rmd" target="_blank">a6c38f8</a>
</td>
<td>
Patron
</td>
<td>
2024-06-18
</td>
<td>
Add home, experiment, and simulation pages
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div
id="empirical-power-and-false-discovery-rate-for-different-target-thresholds"
class="section level2">
<h2>Empirical Power and False Discovery Rate for different target
thresholds</h2>
<p>We take the colon cancer (CRC) dataset which encompasses five
distinct studies each originating from different countries. We
calculated the average proportions of species across all <span
class="math inline">\(574\)</span> samples from five metagenomic studies
and retained the top <span class="math inline">\(p = 100\)</span> most
abundant species. Then, we randomly sampled <span
class="math inline">\(D = 500\)</span> samples from the cohort of <span
class="math inline">\(574\)</span> observations and normalized their raw
OTU counts row-wise, ensuring each row sums to <span
class="math inline">\(1\)</span>. To avoid zero probabilities during
multinomial sampling, a small constant (<span
class="math inline">\(0.5\)</span>) was added to the raw OTU counts
before normalization.</p>
<p>Next, we paired the calculated true proportions <span
class="math inline">\(\Pi\)</span> with sequencing depths resampled from
the original dataset. Each sequencing depth was independently resampled
and paired with a proportion vector from a different subject. Using
these paired proportions and sequencing depths, we generated the taxa
count matrix <span class="math inline">\(\mathbf{X}^{D \times
p}\)</span> via multinomial sampling. For each sample <span
class="math inline">\(d = 1, 2, \ldots, 500\)</span>, the count vector
corresponding to the <span class="math inline">\(d^{\text{th}}\)</span>
row of <span class="math inline">\(\mathbf{X}\)</span> was generated as:
<span class="math display">\[
\mathbf{X}_d \sim \text{Multinomial}(N_d, \Pi_d),
\]</span> where <span class="math inline">\(N_d\)</span> is the
sequencing depth for sample <span class="math inline">\(d\)</span>, and
<span class="math inline">\(\Pi_d\)</span> is the vector of true
proportions for sample <span class="math inline">\(d\)</span> with <span
class="math inline">\(\Pi_{dj}\)</span> (j = 1…p) as its element.</p>
<p>Given the simulated count, among the <span
class="math inline">\(p=100\)</span> taxa, the first <span
class="math inline">\(30\)</span> were assumed to be associated with the
outcome, while the remaining <span class="math inline">\(70\)</span>
taxa had no effect. To simulate outcomes reflecting realistic microbiome
data, signal strengths <span class="math inline">\(s\)</span> for the
top <span class="math inline">\(30\)</span> bio-markers were defined as:
<span class="math display">\[
s_j \overset{d}{=}  
\begin{cases}
    (2Z_j - 1) \frac{U_1}{\sqrt{C_j}}, &amp; \text{for continuous
outcomes}, \\
    (2Z_j - 1) \frac{U_2}{\sqrt{C_j}}, &amp; \text{for binary outcomes},
\end{cases}
\]</span> where <span class="math inline">\(U_1 \sim \text{Uniform}(1.5,
3)\)</span> and <span class="math inline">\(U_2 \sim \text{Uniform}(3,
10)\)</span> represent the signal magnitudes, <span
class="math inline">\(Z_j \sim \text{Bernoulli}(0.5)\)</span> determines
the sign of <span class="math inline">\(s_j\)</span> (<span
class="math inline">\(+1\)</span> or <span
class="math inline">\(-1\)</span>), and <span
class="math inline">\(C_j\)</span> is the column-wise average abundance
of the <span class="math inline">\(j^{\text{th}}\)</span> bio-marker in
<span class="math inline">\(\Pi\)</span>.</p>
<p>Outcomes were generated using the following model: <span
class="math display">\[
   g(\mathbb{E}(y_d)) = \sum_{j=1}^{p} \left \{ \Pi_{dj}^2 \times
\left(\frac{s_j}{2}\right) + \Pi_{dj} \times s_j \right \}
\]</span></p>
<p>where <span class="math inline">\(g(\cdot)\)</span> is the identity
link for continuous outcomes and the logit link for binary outcomes. We
replicate a single iteration of this setting for both continuous and
binary outcomes.</p>
<pre class="r"><code>generate_data &lt;- function(p,seed){
  dcount &lt;- count[,order(decreasing=T,colSums(count,na.rm=T),apply(count,2L,paste,collapse=&#39;&#39;))] ## ordering the columns w/ decreasing abundance
  ####### Randomly sampling patients from 574 observations #######
  set.seed(seed)
  norm_count &lt;- count/rowSums(count)
  col_means &lt;- colMeans(norm_count &gt; 0)
  indices &lt;- which(col_means &gt; 0.2)
  sorted_indices &lt;- indices[order(col_means[indices], decreasing=TRUE)]
  if(p %in% c(100,200,300,400)){
    dcount &lt;- count[,sorted_indices][,1:p]
    sel_index &lt;- sort(sample(1:nrow(dcount), 500))
    dcount &lt;- dcount[sel_index,]
    original_OTU &lt;- dcount + 0.5
    seq_depths &lt;- rowSums(original_OTU)
    Pi = sweep(original_OTU, 1, seq_depths, &quot;/&quot;)
    n = nrow(Pi)
    
    col_abundances = colMeans(Pi)
    
    ##### Generating continuous responses ######
    set.seed(1)
    signals = (2 * rbinom(30, 1, 0.5) - 1) * runif(30, 1.5, 3)
    kBeta = c(signals / sqrt(col_abundances[1:30]), rep(0, p - 30))
    eps=rnorm(n,mean = 0, sd=1)
    Y &lt;- Pi^2 %*% (kBeta/2) + Pi %*% kBeta + eps
    
    ##### Generating binary responses #####
    set.seed(1)
    signals = (2 * rbinom(30, 1, 0.5) - 1) * runif(30, 3, 10)
    kBeta = c(signals / sqrt(col_abundances[1:30]), rep(0, p - 30))
    pr = 1/(1+exp(-(Pi^2 %*% (kBeta/2) + Pi %*% kBeta)))
    Y_bin = rbinom(n,1,pr)
    
    ######### Generate a copy of X #########
    X &lt;- matrix(0, nrow = nrow(Pi), ncol = ncol(Pi))
    nSeq &lt;- seq_depths
    # Loop over each row to generate the new counts based on the multinomial distribution
    
    set.seed(1)
    
    for (i in 1:nrow(Pi)) {
      X[i, ] &lt;- rmultinom(1, size = nSeq, prob = Pi[i, ])
    }  
  } else if (p %in% c(200,300,400)) {
    print(&quot;Enter p within 100 to 400&quot;)
  }
  colnames(X) &lt;- colnames(Pi)
  return(list(Y = Y, X = X, Y_bin = Y_bin, index = 1:30))
}

ntaxa = 100 # Change to p = 200, 300, 400 accordingly.

X &lt;- generate_data(p=ntaxa, seed=1)$X
Y1 &lt;- generate_data(p=ntaxa, seed=1)$Y</code></pre>
<p>On extracting the sample taxa count matrix <span
class="math inline">\(X\)</span> and continuous outcomes <span
class="math inline">\(Y_1\)</span>, we train the zinck model using ADVI.
Plugging in the learnt parameters into the generative framework of
<code>zinck</code>, we generate the knockoff matrix <span
class="math inline">\(\tilde{X}\)</span>.</p>
<pre class="r"><code>fit &lt;- fit.zinck(X,num_clusters = 15,method=&quot;ADVI&quot;,seed=12,elbo_samples = 100)
theta &lt;- fit$theta
beta &lt;- fit$beta
X_tilde &lt;- zinck::generateKnockoff(X,theta,beta,seed=1) ## getting the knockoff copy</code></pre>
<p>Now that we have generated the knockoff copy, we will fit a model
associating the response <span class="math inline">\(Y_1\)</span> with
the augmented set of covariates <span
class="math inline">\([X,\tilde{X}]^{D \times 2p}\)</span>. We fit a
Random Forest model and compute the importance statistics for each
feature. Then, we detect the non-zero features for the FDR threshold of
<span class="math inline">\(0.2\)</span>. Then, we compute the Power
(True Positive Rate) and the empirical FDR by comparing the selected set
of taxa with the true set of taxa indexed from <span
class="math inline">\(1,\dots,30\)</span>.</p>
<pre class="r"><code>index &lt;- 1:30
################### Continuous Outcomes #########################
############## Target FDR = 0.2 ###############
cts_rf &lt;- suppressWarnings(zinck.filter(X,X_tilde,Y1,model=&quot;Random Forest&quot;,fdr=0.2,offset=0,mtry=200,seed=12,metric = &quot;Accuracy&quot;, rftuning = TRUE))
index_est &lt;- cts_rf[[&quot;selected&quot;]]

### Evaluation metrics ###
FN &lt;- sum(index %in% index_est == FALSE) ## False Negatives
FP &lt;- sum(index_est %in% index == FALSE) ## False Positives
TP &lt;- sum(index_est %in% index == TRUE) ## True Positives

estimated_FDR_zinck_cts &lt;- FP / (FP + TP) # Evaluating the empirical False Discovery Rate
estimated_power_zinck_cts &lt;- TP / (TP + FN) # Evaluating the empirical Power or TPR

print(paste(&quot;Estimated FDR for Target 0.2 (Continuous case):&quot;, estimated_FDR_zinck_cts))</code></pre>
<pre><code>[1] &quot;Estimated FDR for Target 0.2 (Continuous case): 0.111111111111111&quot;</code></pre>
<pre class="r"><code>print(paste(&quot;Estimated Power for Target 0.2 (Continuous case):&quot;, estimated_power_zinck_cts))</code></pre>
<pre><code>[1] &quot;Estimated Power for Target 0.2 (Continuous case): 0.533333333333333&quot;</code></pre>
<p>It is to be noted that <code>zinck</code> is outcome agnostic! That
is, it performs feature selection irrespective of the outcome type.
Thus, we demonstrate its power and empirical FDR for binary outcomes
<span class="math inline">\(Y_2\)</span>.</p>
<pre class="r"><code>Y2 &lt;- generate_data(p=ntaxa, seed=1)$Y_bin
index &lt;- 1:30
################### Binary Outcomes #########################
############## Target FDR = 0.2 ###############
bin_rf &lt;- suppressWarnings(zinck.filter(X,X_tilde,as.factor(Y2),model=&quot;Random Forest&quot;,fdr=0.2,offset=0,mtry=14,seed=15,metric=&quot;Accuracy&quot;,rftuning = TRUE))
index_est &lt;- bin_rf[[&quot;selected&quot;]]

### Evaluation metrics ###
FN &lt;- sum(index %in% index_est == FALSE) ## False Negatives
FP &lt;- sum(index_est %in% index == FALSE) ## False Positives
TP &lt;- sum(index_est %in% index == TRUE) ## True Positives

estimated_FDR_zinck_bin &lt;- FP / (FP + TP) # Evaluating the empirical False Discovery Rate
estimated_power_zinck_bin &lt;- TP / (TP + FN) # Evaluating the empirical Power or TPR

print(paste(&quot;Estimated FDR for Target 0.2 (Binary case):&quot;, estimated_FDR_zinck_bin))</code></pre>
<pre><code>[1] &quot;Estimated FDR for Target 0.2 (Binary case): 0.0769230769230769&quot;</code></pre>
<pre class="r"><code>print(paste(&quot;Estimated Power for Target 0.2 (Binary case):&quot;, estimated_power_zinck_bin))</code></pre>
<pre><code>[1] &quot;Estimated Power for Target 0.2 (Binary case): 0.4&quot;</code></pre>
<p>We now compare the performance of <code>zinck</code> with two
standard knockoff filters namely, Model-X Knockoff Filter (MX-KF) and
the standard LDA based knockoff filter (LDA-KF) keeping the same FDR
threshold of <span class="math inline">\(0.2\)</span>. It is to be noted
that for fitting MX-KF, we need to generate the second-order knockoff
copies for the log-normalized version of <span
class="math inline">\(\mathbf{X}\)</span>.</p>
<pre class="r"><code>################ Continuous Outcomes (MX-KF) ##################
set.seed(1)
Xlog = log_normalize(X)
Xlog_tilde = create.second_order(Xlog)
index_est &lt;- zinck.filter(Xlog,Xlog_tilde,Y1,model=&quot;Random Forest&quot;,fdr=0.2,seed=3)$selected

### Evaluation metrics ###
FN &lt;- sum(index %in% index_est == FALSE) ## False Negatives
FP &lt;- sum(index_est %in% index == FALSE) ## False Positives
TP &lt;- sum(index_est %in% index == TRUE) ## True Positives

estimated_FDR_KF_cts &lt;- FP / (FP + TP) # Evaluating the empirical False Discovery Rate
estimated_power_KF_cts &lt;- TP / (TP + FN) # Evaluating the empirical Power or TPR

cat(&quot;LDA-KF (Continuous) metrics\n&quot;)</code></pre>
<pre><code>LDA-KF (Continuous) metrics</code></pre>
<pre class="r"><code>cat(&quot;Estimated FDR for Target 0.2:&quot;, estimated_FDR_KF_cts, &quot;\n&quot;)</code></pre>
<pre><code>Estimated FDR for Target 0.2: 0.125 </code></pre>
<pre class="r"><code>cat(&quot;Estimated Power for Target 0.2:&quot;, estimated_power_KF_cts, &quot;\n&quot;)</code></pre>
<pre><code>Estimated Power for Target 0.2: 0.4666667 </code></pre>
<pre class="r"><code>############### Binary Outcomes (MX-KF) ###################
index_est &lt;- zinck.filter(Xlog,Xlog_tilde,as.factor(Y2),model=&quot;Random Forest&quot;,fdr=0.2,seed=21)$selected

### Evaluation metrics ###
FN &lt;- sum(index %in% index_est == FALSE) ## False Negatives
FP &lt;- sum(index_est %in% index == FALSE) ## False Positives
TP &lt;- sum(index_est %in% index == TRUE) ## True Positives

estimated_FDR_KF_bin &lt;- FP / (FP + TP) # Evaluating the empirical False Discovery Rate
estimated_power_KF_bin &lt;- TP / (TP + FN) # Evaluating the empirical Power or TPR

cat(&quot;LDA-KF (Binary) metrics\n&quot;)</code></pre>
<pre><code>LDA-KF (Binary) metrics</code></pre>
<pre class="r"><code>cat(&quot;Estimated FDR for Target 0.2:&quot;, estimated_FDR_KF_bin, &quot;\n&quot;)</code></pre>
<pre><code>Estimated FDR for Target 0.2: 0.1538462 </code></pre>
<pre class="r"><code>cat(&quot;Estimated Power for Target 0.2:&quot;, estimated_power_KF_bin, &quot;\n&quot;)</code></pre>
<pre><code>Estimated Power for Target 0.2: 0.3666667 </code></pre>
<pre class="r"><code>################ Continuous Outcomes (LDA-KF) ##################
df.LDA = as(as.matrix(X),&quot;dgCMatrix&quot;)
set.seed(1)
vanilla.LDA &lt;- LDA(df.LDA,k=8,method=&quot;VEM&quot;) 
theta.LDA &lt;- vanilla.LDA@gamma
beta.LDA &lt;- vanilla.LDA@beta
beta.LDA &lt;- t(apply(beta.LDA,1,function(row) row/sum(row)))
X_tilde.LDA &lt;- zinck::generateKnockoff(X,theta.LDA,beta.LDA,seed=1) ## Generating vanilla LDA knockoff copy

index_est &lt;- zinck.filter(X,X_tilde.LDA,Y1,model=&quot;Random Forest&quot;,fdr=0.2,offset=0,seed=34)$selected

### Evaluation metrics ###
FN &lt;- sum(index %in% index_est == FALSE) ## False Negatives
FP &lt;- sum(index_est %in% index == FALSE) ## False Positives
TP &lt;- sum(index_est %in% index == TRUE) ## True Positives

estimated_FDR_LDA_cts &lt;- FP / (FP + TP) # Evaluating the empirical False Discovery Rate
estimated_power_LDA_cts &lt;- TP / (TP + FN) # Evaluating the empirical Power or TPR

cat(&quot;LDA-KF (Continuous) metrics\n&quot;)</code></pre>
<pre><code>LDA-KF (Continuous) metrics</code></pre>
<pre class="r"><code>cat(&quot;Estimated FDR for Target 0.2:&quot;, estimated_FDR_LDA_cts, &quot;\n&quot;)</code></pre>
<pre><code>Estimated FDR for Target 0.2: 0 </code></pre>
<pre class="r"><code>cat(&quot;Estimated Power for Target 0.2:&quot;, estimated_power_LDA_cts, &quot;\n&quot;)</code></pre>
<pre><code>Estimated Power for Target 0.2: 0.1333333 </code></pre>
<pre class="r"><code>############### Binary Outcomes (LDA-KF) ###################

index_est &lt;- zinck.filter(X,X_tilde.LDA,as.factor(Y2),model=&quot;Random Forest&quot;,fdr=0.2,offset=0,seed=34)$selected

### Evaluation metrics ###
FN &lt;- sum(index %in% index_est == FALSE) ## False Negatives
FP &lt;- sum(index_est %in% index == FALSE) ## False Positives
TP &lt;- sum(index_est %in% index == TRUE) ## True Positives

estimated_FDR_LDA_bin &lt;- FP / (FP + TP) # Evaluating the empirical False Discovery Rate
estimated_power_LDA_bin &lt;- TP / (TP + FN) # Evaluating the empirical Power or TPR

cat(&quot;LDA-KF (Binary) metrics\n&quot;)</code></pre>
<pre><code>LDA-KF (Binary) metrics</code></pre>
<pre class="r"><code>cat(&quot;Estimated FDR for Target 0.2:&quot;, estimated_FDR_LDA_bin, &quot;\n&quot;)</code></pre>
<pre><code>Estimated FDR for Target 0.2: 0 </code></pre>
<pre class="r"><code>cat(&quot;Estimated Power for Target 0.2:&quot;, estimated_power_LDA_bin, &quot;\n&quot;)</code></pre>
<pre><code>Estimated Power for Target 0.2: 0.3 </code></pre>
<p>To summarize the metrics for the three methods, we create the
following table.</p>
<pre class="r"><code># Example collected results for demonstration purposes
results &lt;- data.frame(
  Method = c(&quot;Zinck&quot;, &quot;MX-KF&quot;, &quot;LDA-KF&quot;, &quot;Zinck&quot;, &quot;MX-KF&quot;, &quot;LDA-KF&quot;),
  Outcome = c(&quot;Continuous&quot;, &quot;Continuous&quot;, &quot;Continuous&quot;, &quot;Binary&quot;, &quot;Binary&quot;, &quot;Binary&quot;),
  Power = c(estimated_power_zinck_cts, estimated_power_KF_cts, estimated_power_LDA_cts, estimated_power_zinck_bin, estimated_power_KF_bin, estimated_power_LDA_bin), 
  FDR = c(estimated_FDR_zinck_cts, estimated_FDR_KF_cts, estimated_FDR_LDA_cts, estimated_FDR_zinck_bin, estimated_FDR_KF_bin, estimated_FDR_LDA_bin)   
)
kable(results, format = &quot;html&quot;, caption = &quot;Power and Empirical FDR Summary&quot;) %&gt;%
  kable_styling(bootstrap_options = c(&quot;striped&quot;, &quot;hover&quot;))</code></pre>
<table class="table table-striped table-hover" style="margin-left: auto; margin-right: auto;">
<caption>
Power and Empirical FDR Summary
</caption>
<thead>
<tr>
<th style="text-align:left;">
Method
</th>
<th style="text-align:left;">
Outcome
</th>
<th style="text-align:right;">
Power
</th>
<th style="text-align:right;">
FDR
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Zinck
</td>
<td style="text-align:left;">
Continuous
</td>
<td style="text-align:right;">
0.5333333
</td>
<td style="text-align:right;">
0.1111111
</td>
</tr>
<tr>
<td style="text-align:left;">
MX-KF
</td>
<td style="text-align:left;">
Continuous
</td>
<td style="text-align:right;">
0.4666667
</td>
<td style="text-align:right;">
0.1250000
</td>
</tr>
<tr>
<td style="text-align:left;">
LDA-KF
</td>
<td style="text-align:left;">
Continuous
</td>
<td style="text-align:right;">
0.1333333
</td>
<td style="text-align:right;">
0.0000000
</td>
</tr>
<tr>
<td style="text-align:left;">
Zinck
</td>
<td style="text-align:left;">
Binary
</td>
<td style="text-align:right;">
0.4000000
</td>
<td style="text-align:right;">
0.0769231
</td>
</tr>
<tr>
<td style="text-align:left;">
MX-KF
</td>
<td style="text-align:left;">
Binary
</td>
<td style="text-align:right;">
0.3666667
</td>
<td style="text-align:right;">
0.1538462
</td>
</tr>
<tr>
<td style="text-align:left;">
LDA-KF
</td>
<td style="text-align:left;">
Binary
</td>
<td style="text-align:right;">
0.3000000
</td>
<td style="text-align:right;">
0.0000000
</td>
</tr>
</tbody>
</table>
<p>This shows that <code>zinck</code> has decent power and controlled
FDR for both binary and continuous outcome scenarios compared to the
other knockoff filters.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.3 (2022-03-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur/Monterey 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] kableExtra_1.4.0     knitr_1.43           topicmodels_0.2-14  
 [4] glmnet_4.1-7         Matrix_1.5-1         dplyr_1.1.2         
 [7] kosel_0.0.1          phyloseq_1.38.0      rstan_2.21.8        
[10] StanHeaders_2.21.0-7 ggplot2_3.4.2        knockoff_0.3.6      
[13] reshape2_1.4.4       zinck_0.0.0.9000    

loaded via a namespace (and not attached):
  [1] colorspace_2.1-0       modeltools_0.2-23      rprojroot_2.0.3       
  [4] XVector_0.34.0         fs_1.6.2               rstudioapi_0.14       
  [7] RSpectra_0.16-1        fansi_1.0.4            ranger_0.15.1         
 [10] xml2_1.3.4             codetools_0.2-19       splines_4.1.3         
 [13] cachem_1.0.8           ade4_1.7-22            jsonlite_1.8.5        
 [16] workflowr_1.7.1        cluster_2.1.4          compiler_4.1.3        
 [19] fastmap_1.1.1          cli_3.6.1              later_1.3.1           
 [22] htmltools_0.5.5        prettyunits_1.1.1      tools_4.1.3           
 [25] igraph_1.4.2           NLP_0.2-1              gtable_0.3.3          
 [28] glue_1.6.2             GenomeInfoDbData_1.2.7 Rcpp_1.0.10           
 [31] slam_0.1-50            Biobase_2.54.0         jquerylib_0.1.4       
 [34] vctrs_0.6.5            Biostrings_2.62.0      rhdf5filters_1.6.0    
 [37] multtest_2.50.0        ape_5.7-1              svglite_2.1.1         
 [40] nlme_3.1-162           iterators_1.0.14       ordinalNet_2.12       
 [43] xfun_0.39              stringr_1.5.0          ps_1.7.5              
 [46] lifecycle_1.0.3        zlibbioc_1.40.0        MASS_7.3-60           
 [49] scales_1.2.1           promises_1.2.0.1       parallel_4.1.3        
 [52] biomformat_1.22.0      rhdf5_2.38.1           inline_0.3.19         
 [55] yaml_2.3.7             gridExtra_2.3          loo_2.6.0             
 [58] sass_0.4.6             stringi_1.7.12         highr_0.10            
 [61] S4Vectors_0.32.4       foreach_1.5.2          randomForest_4.7-1.1  
 [64] permute_0.9-7          BiocGenerics_0.40.0    pkgbuild_1.4.2        
 [67] shape_1.4.6            GenomeInfoDb_1.30.1    rlang_1.1.1           
 [70] pkgconfig_2.0.3        systemfonts_1.0.4      matrixStats_0.63.0    
 [73] bitops_1.0-7           evaluate_0.21          lattice_0.21-8        
 [76] Rhdf5lib_1.16.0        Rdsdp_1.0.5.2          processx_3.8.1        
 [79] tidyselect_1.2.0       plyr_1.8.8             magrittr_2.0.3        
 [82] R6_2.5.1               IRanges_2.28.0         generics_0.1.3        
 [85] DBI_1.1.3              pillar_1.9.0           whisker_0.4.1         
 [88] withr_2.5.0            mgcv_1.8-42            survival_3.5-5        
 [91] RCurl_1.98-1.12        tibble_3.2.1           crayon_1.5.2          
 [94] utf8_1.2.3             rmarkdown_2.22         grid_4.1.3            
 [97] data.table_1.14.8      callr_3.7.3            git2r_0.32.0          
[100] vegan_2.6-4            digest_0.6.31          tm_0.7-8              
[103] httpuv_1.6.11          RcppParallel_5.1.7     stats4_4.1.3          
[106] munsell_0.5.0          viridisLite_0.4.2      bslib_0.5.0           </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
