---
title: "Inflammatory Bowel Disease (IBD) Study"
author: "Soham Ghosh"
date: "2024-06-18"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## IBD Data Analysis 

```{r defn, echo=FALSE}
library(zinck)
library(reshape2)
library(knockoff)
library(ggplot2)
library(rstan)
library(phyloseq)
library(kosel)
zinck.threshold <- function(W, fdr) {
  msel <- ko.sel(W, print = FALSE, method = "gaps")
  T = (1 - fdr) * msel$threshold
  return(T)
}
```


Out of the ten uniformly processed 16S rRNA gene sequencing studies of the IBD mucosal/stool microbiome (https://github.com/biobakery/ibd_paper/tree/paper_publication), we focus on five studies -- RISK (430 cases, 201 controls), CS PRISM (359 cases, 38 controls), HMP2 (59 cases, 22 controls), Pouchitis (308 cases, 45 controls), and Mucosal IBD (36 cases, 47 controls). Here cases indicate patients with Ulcerative Colitis (UC) or Crohnâ€™s Disease (CD). We included all $249$ IBD genera in our analyses.

```{r IBD-data}

################################################################################
###################### IBD data genus level ####################################

load("/Users/Patron/Documents/zinLDA research/genera.RData") ## Loading the meta IBD studies
combined_studies <- as.data.frame(t(physeq_genera@otu_table))
study_names <- physeq_genera@sam_data[["dataset_name"]]

## RISK ##
risk_indices <- which(study_names == "RISK")
risk_otu <- combined_studies[risk_indices, ]

## CS-PRISM ##
prism_indices <- which(study_names == "CS-PRISM")
prism_otu <- combined_studies[prism_indices, ]

## HMP2 ##
HMP_indices <- which(study_names == "HMP2")
hmp_otu <- combined_studies[HMP_indices, ]

## Pouchitis ##
pouchitis_indices <- which(study_names == "Pouchitis")
pouchitis_otu <- combined_studies[pouchitis_indices, ]

## MucosalIBD ##
mi_indices <- which(study_names == "MucosalIBD")
mi_otu <- combined_studies[mi_indices, ]

metadata <- physeq_genera@sam_data

# Get the indices of the last occurrence of each unique subject_accession
latest_indices <- sapply(unique(metadata$subject_accession), function(x) {
  max(which(metadata$subject_accession == x))
})

# Subset the metadata to keep only the latest samples
latest_metadata <- metadata[latest_indices, ]

study_names <- c("CS-PRISM","HMP2","MucosalIBD","Pouchitis","RISK")
meta_IBD <- latest_metadata[latest_metadata$dataset_name %in% study_names,]

# Extract row names for each study
risk_row_names <- row.names(latest_metadata[latest_metadata$dataset_name == "RISK", ])
prism_row_names <- row.names(latest_metadata[latest_metadata$dataset_name == "CS-PRISM", ])
hmp_row_names <- row.names(latest_metadata[latest_metadata$dataset_name == "HMP2", ])
mi_row_names <- row.names(latest_metadata[latest_metadata$dataset_name == "MucosalIBD", ])
pouchitis_row_names <- row.names(latest_metadata[latest_metadata$dataset_name == "Pouchitis", ])

# Subset the OTU matrices and Y vectors based on these row names
risk_otu_latest <- risk_otu[row.names(risk_otu) %in% risk_row_names, ]
prism_otu_latest <- prism_otu[row.names(prism_otu) %in% prism_row_names, ]
hmp_otu_latest <- hmp_otu[row.names(hmp_otu) %in% hmp_row_names, ]
mi_otu_latest <- mi_otu[row.names(mi_otu) %in% mi_row_names, ]
pouchitis_otu_latest <- pouchitis_otu[row.names(pouchitis_otu) %in% pouchitis_row_names, ]

combined_otu <- rbind(prism_otu_latest,hmp_otu_latest,mi_otu_latest,
                      pouchitis_otu_latest,risk_otu_latest)
combined_otu <- combined_otu[ match(rownames(meta_IBD), rownames(combined_otu)), ] 
## To make sure samples in OTU correspond to the samples in meta

IBD_resp <- meta_IBD$disease
Y <- ifelse(IBD_resp %in% c("CD", "UC"), 1, 0)

X <- combined_otu

```

We train the zinck model on $X$ with the optimal number of clusters (27), and use the posterior estimates of the latent parameters to generate the knockoff matrix.

```{r zinck-IBD, message=FALSE, warning=FALSE, results='hide'}
zinck_code <- "data {
  int<lower=1> K; // num topics
  int<lower=1> V; // num words
  int<lower=0> D; // num docs
  int<lower=0> n[D, V]; // word counts for each doc

  // hyperparameters
  vector<lower=0, upper=1>[V] delta;
}

parameters {
  simplex[K] theta[D]; // topic mixtures
  vector<lower=0, upper=1>[V] zeta[K]; // zero-inflated betas
  vector<lower=0>[V] gamma1[K];
  vector<lower=0>[V] gamma2[K];
  vector<lower=0>[K] alpha;
}

transformed parameters {
  vector[V] beta[K];

  // Efficiently compute beta using vectorized operations
  for (k in 1:K) {
    vector[V] cum_log1m;
    cum_log1m[1:(V - 1)] = cumulative_sum(log1m(zeta[k, 1:(V - 1)]));
    cum_log1m[V] = 0;
    beta[k] = zeta[k] .* exp(cum_log1m);
    beta[k] = beta[k] / sum(beta[k]);
  }
}


model {
  for (k in 1:K) {
    alpha[k] ~ gamma(100,100);  // Change these hyperparameters as needed
  }
  for (d in 1:D) {
    theta[d] ~ dirichlet(alpha);
  }
  for (k in 1:K) {
    for (m in 1:V) {
        gamma1[k,m] ~ gamma(1,1);
        gamma2[k,m] ~ gamma(1,1);
    }
  }

  // Zero-inflated beta likelihood and data likelihood
  for (k in 1:K) {
    for (m in 1:V) {
      real lp_non_zero = bernoulli_lpmf(0 | delta[m]) + beta_lpdf(zeta[k, m] | gamma1[k, m], gamma2[k, m]);
      real lp_zero = bernoulli_lpmf(1 | delta[m]);
      target += log_sum_exp(lp_non_zero, lp_zero);
    }
  }

  // Compute the eta values and data likelihood more efficiently
  for (d in 1:D) {
    vector[V] eta = theta[d, 1] * beta[1];
    for (k in 2:K) {
      eta += theta[d, k] * beta[k];
    }
    eta = eta / sum(eta);
    n[d] ~ multinomial(eta);
  }
}
"

stan.model = stan_model(model_code = zinck_code)


dlt <- rep(0,ncol(X)) ## Initializing the deltas with the sparsity of each column

for(t in (1:ncol(X)))
{
  dlt[t] <- 1-mean(X[,t]>0)
  if(dlt[t]==0)
  {
    dlt[t] = dlt[t]+0.01
  }
  if (dlt[t]==1)
  {
    dlt[t] = dlt[t]-0.01
  }
  
}

zinLDA_stan_data <- list(
  K = 27,
  V = ncol(X),
  D = nrow(X),
  n = X,
  delta = dlt
)

## Fitting the zinck model ##
set.seed(1)
fitIBD <- vb(stan.model, data=zinLDA_stan_data, algorithm="meanfield", importance_resampling=TRUE, iter=10000,tol_rel_obj=0.01,elbo_samples=500)

theta <- fitIBD@sim[["est"]][["theta"]]
beta <- fitIBD@sim[["est"]][["beta"]]
X_tilde <- zinck::generateKnockoff(X,theta,beta,seed=1) ## Generating the kncokoff copy


```

Fitting the Random Forest model associating the augmented set of covariates with the outcome of interest, we extract the Feature Importance scores. 

```{r imp-scores, fig.width=10, fig.height=8}

set.seed(2)
W <- stat.random_forest(X,X_tilde,as.factor(Y))
T <- zinck.threshold(W,fdr=0.1)
names_zinck <- colnames(X[,which(W>=T)])

names_zinck <- c("Bilophila genus (f. Desulfovibrionaceae)","Unknown genus (f. Ruminococcaceae)","Holdemania genus (f. Erysipelotrichaceae)","Parabacteroides genus (f. Porphyromonadaceae)","Unknown genus (f. Rikenellaceae)", "Bacteroides genus (f. Bacteroidaceae)","Oscillospira genus (f. Ruminococcaceae)","Phascolarctobacterium genus (f. Veillonellaceae)","Coprococcus genus (f. Lachnospiraceae)","Blautia genus (f. Lachnospiraceae)", "Lachnospira genus (f. Lachnospiraceae)","Faecalibacterium genus (f. Ruminococcaceae)")


Xnorm <- X / rowSums(X)

# Calculate the column sums for cases and controls
case_sums <- colMeans(Xnorm[Y == 1, which(W>=T)])
control_sums <- colMeans(Xnorm[Y == 0, which(W>=T)])

colors <- ifelse(case_sums > control_sums, "red", "blue")


# Create a data frame for plotting
data.genus <- data.frame(
  impscores = sort(W[which(W >= T)], decreasing = FALSE),
  name = factor(names_zinck, levels = names_zinck),
  y = seq(length(names_zinck)) * 0.9,
  color = colors
)


# Create the plot
plt.genus <- ggplot(data.genus) +
  geom_col(aes(impscores, name, fill = color), width = 0.6) +
  scale_fill_identity() +
  theme_bw() +
  ylab("Genera") +
  xlab("Feature Statistic") +
  theme(
    axis.title.x = element_text(size = 22),
    axis.title.y = element_text(size = 22),
    axis.text.x = element_text(size = 18),
    axis.text.y = element_text(size = 18) # Adjust fontface based on the condition
  )
print(plt.genus)


```

