---
title: "Inflammatory Bowel Disease (IBD) Study"
author: "Soham Ghosh"
date: "2024-06-18"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## IBD Data Analysis 

Out of the ten uniformly processed 16S rRNA gene sequencing studies of the IBD mucosal/stool microbiome (https://github.com/biobakery/ibd_paper/tree/paper_publication), we focus on five studies -- RISK (430 cases, 201 controls), CS PRISM (359 cases, 38 controls), HMP2 (59 cases, 22 controls), Pouchitis (308 cases, 45 controls), and Mucosal IBD (36 cases, 47 controls). Here cases indicate patients with Ulcerative Colitis (UC) or Crohnâ€™s Disease (CD). We included all $249$ IBD genera in our analyses.

```{r IBD-data}
library(zinck)
library(reshape2)
library(knockoff)
library(ggplot2)
library(rstan)
library(phyloseq)

################################################################################
###################### IBD data genus level ####################################

load("/Users/Patron/Documents/zinLDA research/genera.RData") ## Loading the meta IBD studies
combined_studies <- as.data.frame(t(physeq_genera@otu_table))
study_names <- physeq_genera@sam_data[["dataset_name"]]

#### Study : RISK ####
risk_indices <- which(study_names == "RISK")
risk_otu <- combined_studies[risk_indices, ]
IBD_resp <- physeq_genera@sam_data[["disease"]][risk_indices]
risk_Y <- ifelse(IBD_resp %in% c("CD", "UC"), 1, 0) ## Labelling "CD" or "UC" as 1, rest as 0

#### Study : CS-PRISM ####
prism_indices <- which(study_names == "CS-PRISM")
prism_otu <- combined_studies[prism_indices, ]
IBD_resp <- physeq_genera@sam_data[["disease"]][prism_indices]
prism_Y <- ifelse(IBD_resp %in% c("CD", "UC"), 1, 0) 

#### Study : HMP2 ######

hmp_indices <- which(study_names == "HMP2")
hmp_otu <- combined_studies[hmp_indices, ]
IBD_resp <- physeq_genera@sam_data[["disease"]][hmp_indices]
hmp_Y <- ifelse(IBD_resp %in% c("CD", "UC"), 1, 0) 

##### Study : MucosalIBD #####

mi_indices <- which(study_names == "MucosalIBD")
mi_otu <- combined_studies[mi_indices, ]
IBD_resp <- physeq_genera@sam_data[["disease"]][mi_indices]
mi_Y <- ifelse(IBD_resp %in% c("CD", "UC"), 1, 0) 

##### Study : Pouchitis #####

pouchitis_indices <- which(study_names == "Pouchitis")
pouchitis_otu <- combined_studies[pouchitis_indices, ]
IBD_resp <- physeq_genera@sam_data[["disease"]][pouchitis_indices]
pouchitis_Y <- ifelse(IBD_resp %in% c("CD", "UC"), 1, 0) 


######### Combining all the 5 studies together #############
X <- rbind(risk_otu,prism_otu,hmp_otu,mi_otu,pouchitis_otu)
Y <- c(risk_Y,prism_Y,hmp_Y,mi_Y,pouchitis_Y)
```

We train the zinck model on $X$ with the optimal number of clusters (27), and use the posterior estimates of the latent parameters to generate the knockoff matrix.

```{r zinck-IBD}
zinck_code <- "data {
  int<lower=1> K; // num topics
  int<lower=1> V; // num words
  int<lower=0> D; // num docs
  int<lower=0> n[D, V]; // word counts for each doc

  // hyperparameters
  vector<lower=0, upper=1>[V] delta;
}

parameters {
  simplex[K] theta[D]; // topic mixtures
  vector<lower=0, upper=1>[V] zeta[K]; // zero-inflated betas
  vector<lower=0>[V] gamma1[K];
  vector<lower=0>[V] gamma2[K];
  vector<lower=0>[K] alpha;
}

transformed parameters {
  vector[V] beta[K];

  // Efficiently compute beta using vectorized operations
  for (k in 1:K) {
    vector[V] cum_log1m;
    cum_log1m[1:(V - 1)] = cumulative_sum(log1m(zeta[k, 1:(V - 1)]));
    cum_log1m[V] = 0;
    beta[k] = zeta[k] .* exp(cum_log1m);
    beta[k] = beta[k] / sum(beta[k]);
  }
}


model {
  for (k in 1:K) {
    alpha[k] ~ gamma(100,100);  // Change these hyperparameters as needed
  }
  for (d in 1:D) {
    theta[d] ~ dirichlet(alpha);
  }
  for (k in 1:K) {
    for (m in 1:V) {
        gamma1[k,m] ~ gamma(1,1);
        gamma2[k,m] ~ gamma(1,1);
    }
  }

  // Zero-inflated beta likelihood and data likelihood
  for (k in 1:K) {
    for (m in 1:V) {
      real lp_non_zero = bernoulli_lpmf(0 | delta[m]) + beta_lpdf(zeta[k, m] | gamma1[k, m], gamma2[k, m]);
      real lp_zero = bernoulli_lpmf(1 | delta[m]);
      target += log_sum_exp(lp_non_zero, lp_zero);
    }
  }

  // Compute the eta values and data likelihood more efficiently
  for (d in 1:D) {
    vector[V] eta = theta[d, 1] * beta[1];
    for (k in 2:K) {
      eta += theta[d, k] * beta[k];
    }
    eta = eta / sum(eta);
    n[d] ~ multinomial(eta);
  }
}
"

stan.model = stan_model(model_code = zinck_code)

X <- X[,order(decreasing=T,colSums(X,na.rm=T),apply(X,2L,paste,collapse=''))] ## ordering the columns w/ decreasing abundance

dlt <- rep(0,ncol(X)) ## Initializing the deltas with the sparsity of each column

for(t in (1:ncol(X)))
{
  dlt[t] <- 1-mean(X[,t]>0)
  if(dlt[t]==0)
  {
    dlt[t] = dlt[t]+0.01
  }
  if (dlt[t]==1)
  {
    dlt[t] = dlt[t]-0.01
  }
  
}

zinLDA_stan_data <- list(
  K = 27,
  V = ncol(X),
  D = nrow(X),
  n = X,
  delta = dlt
)

## Fitting the zinck model ##
set.seed(1)
fitIBD <- vb(stan.model, data=zinLDA_stan_data, algorithm="meanfield", importance_resampling=TRUE, iter=10000,tol_rel_obj=0.01,elbo_samples=500)

theta <- fitIBD@sim[["est"]][["theta"]]
beta <- fitIBD@sim[["est"]][["beta"]]
X_tilde <- zinck::generateKnockoff(X,theta,beta,seed=1) ## Generating the kncokoff copy


```

Fitting the Random Forest model associating the augmented set of covariates with the outcome of interest, we extract the Feature Importance scores. 

```{r imp-scores}
set.seed(5)
W <- stat.random_forest(X,X_tilde,Y)
T <- knockoff.threshold(W,fdr=0.1,offset = 0) ## This is the knockoff filter threshold
print(which(W>=T))

names <- colnames(X[,which(W>=T)]) ## Extracting the names of the important genera
data.genus <- data.frame(
  impscores = sort(W[which(W>=T)], decreasing=FALSE) , 
  name = factor(names, levels = names),
  y = seq(length(names)) * 0.9
)

plt.genus <- ggplot(data.genus) +
  geom_col(aes(impscores, name), fill = "black", width = 0.6)+theme_bw()+ylab("Genera")+xlab("Feature Statistic")

plt.genus

```

